cli/js/signals.ts

```diff
@@ -0,0 +1,171 @@
+   * This method is the shorthand for Deno.signal(Deno.Signal.SIGTERM). */
+  terminate(): SignalStream {
+    return createSignalStream(Signal.SIGTERM);
+  },
+  /** Returns the stream of SIGUSR1 signals.
+   * This method is the shorthand for Deno.signal(Deno.Signal.SIGUSR1). */
+  userDefined1(): SignalStream {
+    return createSignalStream(Signal.SIGUSR1);
+  },
+  /** Returns the stream of SIGUSR2 signals.
+   * This method is the shorthand for Deno.signal(Deno.Signal.SIGUSR2). */
+  userDefined2(): SignalStream {
+    return createSignalStream(Signal.SIGUSR2);
+  },
+  /** Returns the stream of SIGWINCH signals.
+   * This method is the shorthand for Deno.signal(Deno.Signal.SIGWINCH). */
+  windowChange(): SignalStream {
+    return createSignalStream(Signal.SIGWINCH);
+  }
+};
+
+const createSignalStream = (signal: number): SignalStream => {
+  if (build.os === "win") {
+    throw new Error("not implemented!");
+  }
+  return new SignalStream(signal);
+};
+
+const STREAM_DISPOSED_MESSAGE =
+  "No signal is available because signal stream is disposed";
+
+/** SignalStream represents the stream of signals, implements both
+ * AsyncIterator and PromiseLike */
+export class SignalStream implements AsyncIterator<void>, PromiseLike<void> {
+  private rid: number;
+  /** The promise of polling the signal. */
+  private currentPromise: Promise<void> = Promise.resolve();
+  /** The flag, which is true when the stream is disposed. */
+  private disposed = false;
+  constructor(signo: number) {
+    this.rid = sendSync(dispatch.OP_BIND_SIGNAL, { signo }).rid;
+    this.loop();
+  }
+
+  private async pollSignal(): Promise<void> {
+    const { done } = await sendAsync(dispatch.OP_POLL_SIGNAL, {
+      rid: this.rid
+    });
+
+    if (done) {
+      throw new DenoError(ErrorKind.NotFound, STREAM_DISPOSED_MESSAGE);
+    }
+  }
+
+  private async loop(): Promise<void> {
+    while (!this.disposed) {
+      this.currentPromise = this.pollSignal();
+      try {
+        await this.currentPromise;
+      } catch (e) {
+        if (e instanceof DenoError && e.kind === ErrorKind.NotFound) {
+          // If the stream is disposed, then returns silently.
+          return;
+        }
+        // If it's not StreamDisposed error, it's an unexpected error.
+        throw e;
+      }
+    }
+  }
+
+  then<T, S>(
+    f: (v: void) => T | Promise<T>,
+    g?: (v: void) => S | Promise<S>
+  ): Promise<T | S> {
+    return this.currentPromise.then(f, g);
+  }
+
+  async next(): Promise<IteratorResult<void>> {
+    try {
+      await this.currentPromise;
+      return { done: false, value: undefined };
+    } catch (e) {
+      if (e instanceof DenoError && e.kind === ErrorKind.NotFound) {
+        return { done: true, value: undefined };
+      }
+      throw e;
+    }
+  }
+
+  [Symbol.asyncIterator](): AsyncIterator<void> {
+    return this;
+  }
+
+  dispose(): void {
+    this.disposed = true;
+    sendSync(dispatch.OP_UNBIND_SIGNAL, { rid: this.rid });
+  }
```

Comment:
Maybe guard against additional calls to `dispose()`

Reply:
Sounds a good idea.